### ISSUE TITLE
Add support for history@3.0.0

### ISSUE THREAD

**Issue by gajus** [reactjs/react-router#3515]
Related to https://github.com/reactjs/react-router-redux/issues/385.

**Comment by taion** [comment]
I don't think we can (easily) or should try to do this in a backward-compatible way BTW.

**Comment by julien-f** [comment]
Then maybe it should ship with react-router 3.0.0?

**Comment by gajus** [comment]
Delaying 3.0.0 release to include history@3 would make sense. Otherwise you'd have multiple major releases in relatively short time span.

**Comment by taion** [comment]
Ideally yeah. The user-facing API shouldn't change too much in any case.

**Comment by taion** [comment]
Not sure about that – in the v2 history change there was actually a discrete visible API change to the user whenever custom histories were involved (the mandatory user-called `useRouterHistory`).

For going to v3, the user-facing API shouldn't change in most cases – the user only needs to upgrade both libraries.

**Comment by deser** [comment]
Guys,
when it's planned to release react-router 3.0 ?

**Comment by timdorr** [comment]
Apparently the next version [isn't going to use `history`](https://twitter.com/ryanflorence/status/743115221573591040), so this is all moot. @mjackson says [it should be out in a week or so](https://github.com/javascriptair/site/issues/110#issuecomment-226383000). 

Closing this in light of that news.

**Comment by taion** [comment]
Huh? Is that a good idea?

**Comment by timdorr** [comment]
I have no idea. No one was informed they were going off to do their own thing. That seems counter to the opening statements of #3289, but whatever...

**Comment by taion** [comment]
Hmm, I do think the API boundary between React Router and history has room for improvement, and the current situation is actually fairly confusing for users – see this issue and issues w/using e.g. `use-named-routes` with releases.

I do think it's useful to decouple history management from the router somewhat, but the current library pattern does have fixable flaws.

Looking forward to seeing it!

**Comment by timdorr** [comment]
Sure, I'm just whining about not being included in the cool kids club. 😄 

I'm also for some further clarification because of confusion surrounding router state not being available in `react-router-redux` and that really being a `history` integration library (despite the name). It's also pretty confusing for users to have to learn and configure this entirely separate library if they want to do custom things (like named routes). It would be a bit like making your favorite dish with a different ingredient and having to learn how to farm in order to do it. It makes the API less fun to work with. So, hopefully this is going to be a more fun API 👍

**Comment by taion** [comment]
Just thinking about this a little, if I were doing a rewrite of history state management, I would:

1. Rewrite history to follow the new RN navigation paradigm, where the history is essentially just a reducer, written in such a way that it can plug into Redux (but also comes with a default store) – this makes Redux integration "just work" and better follows modern best practices for state management
2. Break up `<Router />` into something like `<LocationProvider><RouteMatcher /></LocationProvider>`, and have the `location` injected down as a prop (so `<LocationProvider>` can be replaced with just the React Redux `<Provider>` if relevant), instead of having the same piece doing router state management also handle subscribing to the history.
3. Customizable route path matching engine (maybe route-by-route)

Hmm... if this rewrite gets stalled, I might actually try to move things in that direction.

**Comment by timdorr** [comment]
Number 2 and 3 are pretty close to what Ryan and Michael are working on (Ryan showed me what they have so far on Friday). I don't think anything is going to stall, since they're pretty excited about it (I am too!). They'll loop us in soon.

**Comment by taion** [comment]
Sure – I'd say that once you factor out a location provider component, then that component really _should_ use history or a similar abstraction if practical.

**Comment by timdorr** [comment]
Oh yeah, it still does use `history`. I misread some things.

**Comment by taion** [comment]
@gaearon recommended that we cut a v3 release with a relative small changeset ahead of the pending rewrite to give users a path forward, with improvements from v2. I'm re-opening this issue and adding it to the v3 milestone.

This is blocked on https://github.com/ReactJSTraining/history/pull/316, and also on the question of how to handle getting the initial location – i.e. should we do a `history.replace(currentLocation)` in the router to ensure that the initial location always has a key on it (when e.g. using browser history)?

cc @mjackson

**Comment by deser** [comment]
```
As Breakage: history.listen no longer calls the callback synchronously once. maybe it would be possible to make setRouteLeaveHook asynchronous?
```
So let me repeat my question

**Comment by timdorr** [comment]
No, the `listen` change has no bearing on `setRouteLeaveHook`. It simply changes the first call behavior.

**Comment by deser** [comment]
:(

**Comment by taion** [comment]
There is no plan right now to make leave hooks async. If you want more complex confirmation behavior, take a look at extending the behavior of history's navigation confirmation hooks to better work with custom confirmation dialogs.

**Comment by taion** [comment]
Folded into https://github.com/reactjs/react-router/issues/3611.

**Comment by taion** [comment]
https://github.com/reactjs/react-router/pull/3647
